File: src/main/java/com/deepseek/model/WorldBook.java
package com.deepseek.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents a World Book (Lorebook) containing multiple entries.
 */
public class WorldBook implements Serializable {
    private static final long serialVersionUID = 1L;
    
    private String id;
    private String name;
    private String description;
    private List<WorldBookEntry> entries;
    private long createdAt;
    private long updatedAt;
    
    public WorldBook() {
        this.id = java.util.UUID.randomUUID().toString();
        this.createdAt = System.currentTimeMillis();
        this.updatedAt = this.createdAt;
        this.entries = new ArrayList<>();
    }
    
    public WorldBook(String name) {
        this();
        this.name = name;
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public List<WorldBookEntry> getEntries() { return entries; }
    public void setEntries(List<WorldBookEntry> entries) { this.entries = entries; }
    
    public long getCreatedAt() { return createdAt; }
    public void setCreatedAt(long createdAt) { this.createdAt = createdAt; }
    
    public long getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(long updatedAt) { this.updatedAt = updatedAt; }
    
    /**
     * Add an entry to this world book.
     */
    public void addEntry(WorldBookEntry entry) {
        if (entry != null) {
            entries.add(entry);
            this.updatedAt = System.currentTimeMillis();
        }
    }
    
    /**
     * Remove an entry by ID.
     */
    public boolean removeEntry(String entryId) {
        boolean removed = entries.removeIf(e -> e.getId().equals(entryId));
        if (removed) {
            this.updatedAt = System.currentTimeMillis();
        }
        return removed;
    }
    
    /**
     * Find all enabled entries that match the given text.
     */
    public List<WorldBookEntry> findMatchingEntries(String text) {
        List<WorldBookEntry> matching = new ArrayList<>();
        for (WorldBookEntry entry : entries) {
            if (entry.matches(text)) {
                matching.add(entry);
            }
        }
        // Sort by priority (higher first)
        matching.sort((a, b) -> Integer.compare(b.getPriority(), a.getPriority()));
        return matching;
    }
    
    @Override
    public String toString() {
        return name != null ? name : "Unnamed World Book";
    }
}

File: cpp_deepseek/src/model/WorldBook.cpp
/**
 * @file WorldBook.cpp
 * @brief Implementation of the WorldBook class.
 */

#include "deepseek/model/WorldBook.hpp"
#include <chrono>
#include <algorithm>
#include <sstream>

namespace deepseek {
namespace model {

namespace {
std::string generateUuid() {
    static std::random_device rd;
    static std::mt19937_64 gen(rd());
    static std::uniform_int_distribution<uint64_t> dis;
    
    uint64_t uuid = dis(gen);
    std::stringstream ss;
    ss << std::hex << std::setfill('0');
    ss << std::setw(8) << (uuid >> 32) << "-";
    ss << std::setw(4) << ((uuid >> 16) & 0xFFFF) << "-";
    ss << std::setw(4) << (uuid & 0xFFFF) << "-";
    ss << std::setw(4) << (dis(gen) >> 48) << "-";
    ss << std::setw(12) << dis(gen);
    return ss.str();
}
} // anonymous namespace

WorldBook::WorldBook()
    : id_(generateUuid())
    , createdAt_(std::chrono::duration_cast<std::chrono::milliseconds>(
          std::chrono::system_clock::now().time_since_epoch()).count())
    , updatedAt_(createdAt_) {
}

WorldBook::WorldBook(const std::string& name)
    : WorldBook() {
    name_ = name;
}

void WorldBook::addEntry(const WorldBookEntry& entry) {
    entries_.push_back(entry);
    updatedAt_ = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
}

bool WorldBook::removeEntry(const std::string& entryId) {
    auto it = std::remove_if(entries_.begin(), entries_.end(),
        [&entryId](const WorldBookEntry& e) { return e.getId() == entryId; });
    
    bool removed = (it != entries_.end());
    if (removed) {
        entries_.erase(it, entries_.end());
        updatedAt_ = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count();
    }
    return removed;
}

std::vector<WorldBookEntry> WorldBook::findMatchingEntries(const std::string& text) const {
    std::vector<WorldBookEntry> matching;
    
    for (const auto& entry : entries_) {
        if (entry.matches(text)) {
            matching.push_back(entry);
        }
    }
    
    // Sort by priority (higher first)
    std::sort(matching.begin(), matching.end(),
        [](const WorldBookEntry& a, const WorldBookEntry& b) {
            return a.getPriority() > b.getPriority();
        });
    
    return matching;
}

std::string WorldBook::toString() const {
    return name_.empty() ? "Unnamed World Book" : name_;
}

} // namespace model
} // namespace deepseek
